File: R/kriging.R

#' Function to create a gstlearn object containing the parameters of 
#' the neighborhood used to perform kriging
#'
#' @keywords internal
#'
.createneigh<-function(neigh){
  if (length(neigh) == 1){
    if (neigh == "unique"){
      Neigh = NeighUnique()
    }
    else if ((sum(is.numeric(neigh))== 1)){
      Neigh =NeighMoving_create(radius = neigh)
    }
  }
  else if ((sum(is.numeric(neigh))== 1) & (length(neigh) == 3)){
    if ((neigh[1] <= neigh[2]) & (neigh[3]>0)){
      Neigh =NeighMoving_create(nmini=neigh[1], nmaxi=neigh[2], radius=neigh[3])
    }
  }
  else if ((sum(is.numeric(neigh))== 1) & (length(neigh) == 2)){
    if (neigh[1] <= neigh[2]){
      Neigh =NeighMoving_create(nmini=neigh[1], nmaxi=neigh[2])
    }
  }
  else{
    stop("The neighborhood definition is not valid")
  }
  return(Neigh)
}

#' Kriging predictions and cross-validation
#'
#' Functions to compute the kriging predictions (`minikriging`) or cross-validations (`minixvalid`).
#'
#' @param dbin Db object containing the data points.
#' @param dbout Db object containing the target points (ignored for `minixvalid`), i.e. the points where the kriging predictor will be computed.
#' @param vname Name of the variable(s) to be predicted, stored as a (vector of) string(s).
#' @param model Model object containing the model used for kriging. For universal kriging, the model should be fitted on residuals.
#' @param type Type of kriging: either "simple" or "ordinary". Ignored if `polDrift` or `extDrift` is specified, in which case, universal kriging is performed.
#' @param polDrift Integer specifying the order of the polynomial drift for universal kriging. 
#' @param extDrift Name of the variable(s) specifying the external drift, stored as a (vector of) string(s), for universal kriging with external drift.
#' @param mean Mean value if simple kriging is requested. Ignored when other types of kriging are performed.
#' @param neighborhood Type of neighborhood to used, either "unique" or a vector of the form (radius), (nmin, nmax) or (nmin, nmax, radius). See details. 
#' @param std Boolean value indicating whether the kriging standard deviations (for the `minikriging` function) or standardized cross-validation errors (for the `minixvalid` function) should be computed.
#' @param prefix Prefix used to store the results
#' 
#' @details The function `minikriging` computes kriging predictions on target locations specified in a Db object (argument `dbout`) using data specified on another Db object (argument `dbin`).
#' 
#' The function `minixvalid` computes kriging 1-fold cross-validation errors on a set of points  specified in a Db object (argument `dbin`): For each point x_i, the function computes the kriging prediction at x_i when using all the other points in Db. 
#' The cross-validation error at x_i is then defined as the difference between this prediction and the actual observed value at x_i. 
#' And the standardized cross-validation error at x_i is defined as the ratio between the cross-validation error at x_i  and the kriging standard-deviation at x_i. 
#' 
#' When specifying moving neighborhoods, the vector (nmin, nmax, radius) indicates the minimum and the maximum number of points in the neighborhood and the radius of the neighborhood.
#' 
#' Finally, note that when adding external drifts, a constant drift (bias term) is automatically added as well.
#' 
#' @return The function `minikriging` directly adds the kriging predictions (and if applicable, standard-deviations) to the Db in `dbout`. The names of these newly created variables will be of the form `prefix.vname.estim` for the predictions and  `prefix.vname.stdev` for the standard deviations.
#' 
#' The function `minixvalid` directly adds the kriging cross-validation errors (and if applicable, standard-deviations) to the Db in `dbin`. The names of these newly created variables will be of the form `prefix.vname.esterr` for the cross-vaidation errors and  `prefix.vname.stderr` for the standardardized errors
#'
#' @export
#'
#' @examples
#' library(minigst)
#'
#' # Load Data
#' data("Scotland") # Dataframe containing the observations
#' data("ScotlandGrid") # Dattaframe containing the target prediction locations
#'
#' # Create Observation grid
#' obsDb=dfToDb(df=Scotland, coordnames=c("Longitude", "Latitude"))
#' 
#'
#' # Create directional experimental variograms along the directions 30 deg and -30 deg for the variable "Elevation"
#' varioExp = vario_exp(db=obsDb, vname="Elevation", dir=c(30,-30), nlag=20, dlag=10.)
#' # Fit a model on the resulting experimental variogram
#' struct_names = c("NUGGET","SPHERICAL", "SPHERICAL")
#' model = model_fit(varioExp, struct=struct_names, pruneModel = TRUE)
#' 
#' # Create target grid
#' targetDb=dfToDbGrid(df=ScotlandGrid,coordnames=c("Longitude","Latitude"))
#' 
#' # Perform ordinary kriging predictions and compute standard deviations for the variable `January_temp`
#' minikriging(obsDb, targetDb, vname = "January_temp", model = model, std = TRUE)
#'  
#'# Plot kriging predictions
#' dbplot_grid(targetDb,color="K.January_temp.estim")
#' dbplot_point(obsDb,size="January_temp",sizeRange=c(0.1,2),add=TRUE,title="Kriging predictions on grid")
#' 
#'# Plot kriging standard deviations
#' dbplot_grid(targetDb,color="K.January_temp.stdev")
#' dbplot_point(obsDb,size="January_temp",sizeRange=c(0.1,2),add=TRUE,title="Kriging std dev")
#' 
minikriging<-function (dbin, dbout, vname, model, type = "ordinary", polDrift = NULL, extDrift=NULL, mean = NULL, neighborhood = "unique", std = TRUE,prefix="K"){
  setVar(dbin,vname)
  Neigh = .createneigh(neighborhood)
  
  if (!is.null(extDrift) || !is.null(polDrift)){
    ## Create a copy of the model without any drift and set the mean
    mdl=Model(model)
    err = mdl$delAllDrifts()
    ## Add drifts to model
    .addDriftsToModel(mdl,polDrift,length(extDrift))
    
    ## Set drifts in databases
    setVar(dbin,extDrift,"Drift")
    setVar(dbout,extDrift,"Drift")
    
    ## Delete previous kriging result
    .deleteExistingVar(dbout,paste0(c(prefix,vname,"estim"),collapse = "."))
    if(std){
      .deleteExistingVar(dbout,paste0(c(prefix,vname,"stdev"),collapse = "."))
    }
    ## Kriging
    err = kriging(dbin=dbin, dbout=dbout, model=mdl, 
                  neigh=Neigh,
                  flag_est=TRUE, flag_std=std, flag_varz=FALSE,
                  namconv=NamingConvention(prefix)
    )
    ## Remove locators automatically assigned by the `kriging` function
    Db_clearLocators(dbout,ELoc_Z())
    Db_clearLocators(dbout,ELoc_F())
    Db_clearLocators(dbin,ELoc_Z())
    Db_clearLocators(dbin,ELoc_F())
  } 
  else if (type == "simple"){
    if (!is.numeric(mean)){stop('The mean is not properly specified for the simple kriging.')}
    
    ## Create a copy of the model without any drift and set the mean
    mdl=Model(model)
    err = mdl$delAllDrifts()
    err = mdl$setMean(mean)
    
    ## Delete previous kriging result
    .deleteExistingVar(dbout,paste0(c(prefix,vname,"estim"),collapse = "."))
    if(std){
      .deleteExistingVar(dbout,paste0(c(prefix,vname,"stdev"),collapse = "."))
    }
    ## Compute kriging
    err = kriging(dbin=dbin, dbout=dbout, model=mdl, 
                  neigh=Neigh,
                  flag_est=TRUE, flag_std=std, flag_varz=FALSE,
                  namconv=NamingConvention(prefix)
    )
    ## Remove locators automatically assigned by the `kriging` function
    Db_clearLocators(dbout,ELoc_Z())
    Db_clearLocators(dbin,ELoc_Z())
    
  }
  else if (type == "ordinary"){
    
    ## Create a copy of the model without any drift and set universality condition
    mdl=Model(model)
    err = mdl$delAllDrifts()
    err = mdl$addDrift(DriftM())
    
    ## Delete previous kriging result
    .deleteExistingVar(dbout,paste0(c(prefix,vname,"estim"),collapse = "."))
    if(std){
      .deleteExistingVar(dbout,paste0(c(prefix,vname,"stdev"),collapse = "."))
    }
    ## Kriging
    err = kriging(dbin=dbin, dbout=dbout, model=mdl, 
                  neigh=Neigh,
                  flag_est=TRUE, flag_std=std, flag_varz=FALSE,
                  namconv=NamingConvention(prefix)
    )
    ## Remove locators automatically assigned by the `kriging` function
    Db_clearLocators(dbout,ELoc_Z())
    Db_clearLocators(dbin,ELoc_Z())
    
    } 
  else {stop("The kriging type should be either simple ordinary, or universal (with a drift properly defined)")}
  
  return(invisible(NULL))
}


#' #' Compute the cross validation error
#' #'
#' #' Function to perform the (1-fold) cross validation for a given kriging exercise.
#' #'
#' #' @param dbin Db containing the data.
#' #' @param vname Name of the variable, stored as a string.
#' #' @param model A \pkg{gstlearn} Model object containing a fitted model.
#' #' @param type Type of kriging, either "simple" or "ordinary".
#' #' @param mean Mean value if simple kriging is requested 
#' #' @param neighborhood Type of neighborhood to used, either "unique" or a vector of the form (radius), (nmin, nmax) or (nmin, nmax, radius) see details. 
#' #' @param std Boolean value indicating whether the prediction standard deviation should be computed.
#' #'
#' #' @details For each observation location x_i, compute the kriging prediction at x_i when using all the observations except the one located at x_i. 
#' #' The cross-validation error at x_i is then defined as the difference between this prediction and the actual observed value at x_i.
#' #' 
#' #' @return a db containing the same variables as the input db + two new variables: 
#' #' One variable (with name ending with .esterr) containing the cross-validation errors, 
#' #' and one variable (with name ending with .stderr) containing the cross-validation errors normalized by the kriging standard deviation.
#' #'
#' #' @export
#' #'
#' #' @examples
#' #' library(minigst)
#' #'
#' #' #todo
#' #' 
#' #' 
#' #'  
#' #'
#' xvalidation <- function (dbin, vname, drift = NULL, model, type = "ordinary",prefix = "OK", mean = NULL, neighborhood = "unique", std = TRUE){
#'   setVar(dbin,vname)
#'   Neigh = .createneigh(neighborhood)
#'   if (type == "simple"){
#'     if (!as.numeric(mean)){stop('The mean is not properly specified for the simple kriging.')}
#'     err = model$setMean(mean)
#'     err = xvalid(db=dbin, model=model, 
#'                  neigh=Neigh,
#'                  flag_xvalid_est=1, flag_xvalid_std=1,  
#'                  namconv=NamingConvention_create(paste0("Xvalid_",prefix), flag_locator = FALSE)
#'     )
#'   }
#'   else if (type == "ordinary"){
#'     err = model$delAllDrifts()
#'     err = model$addDrift(DriftM())
#'     err = xvalid(db=dbin, model=model, 
#'                  neigh=Neigh,
#'                  flag_xvalid_est=1, flag_xvalid_std=1,  
#'                  namconv=NamingConvention_create(paste0("Xvalid_",prefix), flag_locator = FALSE)
#'     )
#'   }
#'   else if (type == "universal" & !is.null(drift)){
#'     for (xvar in drift){
#'       dbin$setLocator(xvar,ELoc_F())
#'     }
#'     err = xvalid(db=dbin, model=model, 
#'                  neigh=Neigh,
#'                  flag_xvalid_est=1, flag_xvalid_std=1,  
#'                  namconv=NamingConvention_create(paste0("Xvalid_",prefix), flag_locator = FALSE)
#'     )
#'   }
#'   else {stop("The kriging type should be either simple, ordinary, or universal")}
#'   return(dbin)
#' }



#' @rdname minikriging
#'
#' @export
#'
minixvalid<-function (dbin, vname, model, type = "ordinary", polDrift = NULL, extDrift=NULL, mean = NULL, neighborhood = "unique", std = TRUE,prefix="Xvalid"){
  setVar(dbin,vname)
  Neigh = .createneigh(neighborhood)
  
  if (!is.null(extDrift) || !is.null(polDrift)){
    ## Create a copy of the model without any drift and set the mean
    mdl=Model(model)
    err = mdl$delAllDrifts()
    ## Add drifts to model
    .addDriftsToModel(mdl,polDrift,length(extDrift))
    
    ## Set drifts in databases
    setVar(dbin,extDrift,"Drift")
    
    ## Delete previous xvalid result
    .deleteExistingVar(dbin,paste0(c(prefix,vname,"esterr"),collapse = "."))
    if(std){
      .deleteExistingVar(dbin,paste0(c(prefix,vname,"stderr"),collapse = "."))
    }
    ## xvalid
    err = xvalid(dbin, model=mdl, 
                 neigh=Neigh,
                 flag_xvalid_est=TRUE, flag_xvalid_std=std,
                 namconv=NamingConvention(prefix)
    )
    ## Remove locators automatically assigned by the `xvalid` function
    Db_clearLocators(dbin,ELoc_Z())
    Db_clearLocators(dbin,ELoc_F())
  }
  else if (type == "simple"){
    if (!is.numeric(mean)){stop('The mean is not properly specified for the simple xvalid.')}
    
    ## Create a copy of the model without any drift and set the mean
    mdl=Model(model)
    err = mdl$delAllDrifts()
    err = mdl$setMean(mean)
    
    ## Delete previous xvalid result
    .deleteExistingVar(dbin,paste0(c(prefix,vname,"esterr"),collapse = "."))
    if(std){
      .deleteExistingVar(dbin,paste0(c(prefix,vname,"stderr"),collapse = "."))
    }
    ## Compute xvalid
    err = xvalid(dbin, model=mdl, 
                  neigh=Neigh,
                  flag_xvalid_est=TRUE, flag_xvalid_std=std, 
                  namconv=NamingConvention(prefix)
    )
    ## Remove locators automatically assigned by the `xvalid` function
    Db_clearLocators(dbin,ELoc_Z())
    
  }
  else if (type == "ordinary"){
    
    ## Create a copy of the model without any drift and set universality condition
    mdl=Model(model)
    err = mdl$delAllDrifts()
    err = mdl$addDrift(DriftM())
    
    ## Delete previous xvalid result
    .deleteExistingVar(dbin,paste0(c(prefix,vname,"esterr"),collapse = "."))
    if(std){
      .deleteExistingVar(dbin,paste0(c(prefix,vname,"stderr"),collapse = "."))
    }
    ## xvalid
    err = xvalid(dbin, model=mdl, 
                  neigh=Neigh,
                  flag_xvalid_est=TRUE, flag_xvalid_std=std,
                  namconv=NamingConvention(prefix)
    )
    ## Remove locators automatically assigned by the `xvalid` function
    Db_clearLocators(dbin,ELoc_Z())
    
  }
  else {stop("The xvalid type should be either simple ordinary, or universal (with a drift properly defined)")}
  
  return(invisible(NULL))
}

File: python/minigst/kriging.py
"""Kriging functions for minigst package."""

import numpy as np
import gstlearn as gl
import pandas as pd
from .model import add_drifts_to_model
from .db import df_to_db


def _create_neigh(neigh):
    """
    Create a gstlearn neighborhood object.

    Args:
        neigh: Neighborhood specification (string or numeric)

    Returns:
        gstlearn Neigh object
    """
    if isinstance(neigh, str):
        if neigh == "unique":
            return gl.NeighUnique.create()
        else:
            raise ValueError(
                "Invalid neighborhood string. Use 'unique' or provide numeric parameters."
            )

    if isinstance(neigh, (int, float)):
        # Single radius
        return gl.NeighMoving.create(radius=float(neigh))

    if isinstance(neigh, (list, tuple, np.ndarray)):
        if len(neigh) == 1:
            # Single radius
            return gl.NeighMoving.create(radius=float(neigh[0]))
        elif len(neigh) == 2:
            # nmin, nmax
            return gl.NeighMoving.create(nmini=int(neigh[0]), nmaxi=int(neigh[1]))
        elif len(neigh) == 3:
            # nmin, nmax, radius
            return gl.NeighMoving.create(
                nmini=int(neigh[0]), nmaxi=int(neigh[1]), radius=float(neigh[2])
            )
        else:
            raise ValueError("Invalid neighborhood specification.")

    raise ValueError("The neighborhood definition is not valid.")


def minikriging(
    dbin,
    dbout,
    vname,
    model,
    type="ordinary",
    pol_drift=None,
    ext_drift=None,
    mean=None,
    nmini=None,
    nmaxi=None,
    radius=None,
    std=True,
    prefix="K",
):
    """
    Compute kriging predictions.

    Args:
        dbin: gstlearn Db object with observation data
        dbout: gstlearn Db object for target predictions
        vname: Variable name(s) to predict (string or list)
        model: gstlearn Model object
        type: Kriging type ('simple' or 'ordinary')
        pol_drift: Polynomial drift order for universal kriging
        ext_drift: External drift variable name(s) for universal kriging
        mean: Mean value for simple kriging
        nmini: Minimum number of points in the neighborood
        nmaxi: Maximum number of points in the neighborood
        radius: Maximum distance between target and data for the neighborood
        std: Boolean, compute standard deviations
        prefix: Prefix for output variable names

    Returns:
        None (results added directly to dbout)

    Details:
        If nmini = nmaxi = radius = None, unique neighborood is used.
    Examples:
        >>> import minigst as mg
        >>> mg.minikriging(obs_db, target_db, vname='temperature',
        ...                model=model, type='ordinary', std=True)
    """

    # Set variable as Z locator
    if isinstance(vname, str):
        vname = [vname]

    # Remove previuous runs with same name
    vdel = [prefix + "." + v + "." + "estim" for v in vname]
    vdel = vdel + [prefix + "." + v + "." + "stdev" for v in vname]
    dbout.deleteColumns(vdel)

    # Set Locators
    dbin.setLocators(vname, gl.ELoc.Z)

    # Create neighborhood
    if nmini is None and nmaxi is None and radius is None:
        neigh = gl.NeighUnique()
    else:
        if nmini is None:
            nmini = 0
        if nmaxi is None:
            nmaxi = dbin.getNSample()
        if radius is None:
            radius = np.inf
        neigh = gl.NeighMoving.create(nmini=nmini, nmaxi=nmaxi, radius=radius)

    # Handle drifts
    dbin.clearLocators(gl.ELoc.F)
    dbout.clearLocators(gl.ELoc.F)
    model_copy = model.clone()

    if ext_drift is not None:
        if isinstance(ext_drift, str):
            ext_drift = [ext_drift]

        # Set external drifts
        dbin.setLocators(ext_drift, gl.ELoc.F)
        dbout.setLocators(ext_drift, gl.ELoc.F)

    ndrifts = dbin.getNLoc(gl.ELoc.F)
    add_drifts_to_model(model_copy, pol_drift, ndrifts, type)

    if type == "simple":
        # Simple kriging
        if mean is None:
            raise ValueError("Mean must be specified for simple kriging.")
        model_copy.setMean(mean)

    err = gl.kriging(
        dbin,
        dbout,
        model_copy,
        neigh,
        flag_est=True,
        flag_std=std,
        namconv=gl.NamingConvention(prefix),
    )


def regression(db, vname, model, pol_drift=None, ext_drift=None, reml=False):
    """
    Compute and display linear regression coefficients (BLUE).

    Args:
        db: gstlearn Db object with observation data
        vname: Variable name (monovariable only)
        model: gstlearn Model object
        pol_drift: Polynomial drift order
        ext_drift: External drift variable name(s)
        reml: Boolean indicating if coefficients have to be computed with restricted maximum likelihood
    Returns:
        A list with
          - the mean parameter
          - the vector of parameters for the polynomial coefficients
          - the vector of parameters for the external drift coefficients
    Examples:
        >>> import minigst as mg
        >>> mg.minikriging(obs_db, vname='temperature',
        ...                model=model)
    """
    if not isinstance(vname, str):
        raise ValueError("You have to specify a single vname")
    db.setLocator(vname, gl.ELoc.Z, cleanSameLocator=True)
    db.clearLocators(gl.ELoc.F)
    if ext_drift is not None:
        if isinstance(ext_drift, str):
            ext_drift = [ext_drift]
    db.setLocators(ext_drift, gl.ELoc.F)
    if pol_drift is None:
        pol_drift = 0
    ndrifts = db.getNLoc(gl.ELoc.F)
    model_copy = model.clone()
    model_copy.setDriftIRF(pol_drift, ndrifts)
    a = gl.Likelihood.createForOptim(model_copy, db, reml)
    a.computeLogLikelihood()
    beta = a.getBeta()
    d = model_copy.getDriftList()
    print("Intercept:", np.round(beta[0], 5))
    for i in range(1, len(beta) - db.getNLoc(gl.ELoc.F)):
        print(d.getDrift(i), ":", np.round(beta[i], 5))
    s = 0
    for i in range(len(beta) - db.getNLoc(gl.ELoc.F), len(beta)):
        print(db.getNameByLocator(gl.ELoc.F, s), ":", np.round(beta[i], 5))
        s += 1


def minixvalid(
    dbin,
    vname,
    model,
    type="ordinary",
    pol_drift=None,
    ext_drift=None,
    mean=None,
    neighborhood="unique",
    std=True,
    prefix="Xvalid",
):
    """
    Compute kriging cross-validation.

    Args:
        dbin: gstlearn Db object with observation data
        vname: Variable name(s) to validate (string or list)
        model: gstlearn Model object
        type: Kriging type ('simple' or 'ordinary')
        pol_drift: Polynomial drift order for universal kriging
        ext_drift: External drift variable name(s) for universal kriging
        mean: Mean value for simple kriging
        neighborhood: Neighborhood specification ('unique' or numeric)
        std: Boolean, compute standardized errors
        prefix: Prefix for output variable names

    Returns:
        None (results added directly to dbin)

    Examples:
        >>> import minigst as mg
        >>> mg.minixvalid(obs_db, vname='temperature', model=model,
        ...               type='ordinary', std=True)
    """
    # Set variable as Z locator
    if isinstance(vname, str):
        vname = [vname]

    # Remove previuous runs with same name
    vdel = [prefix + "." + v + "." + "esterr" for v in vname]
    vdel = vdel + [prefix + "." + v + "." + "stderr" for v in vname]
    dbin.deleteColumns(vdel)

    for vn in vname:
        dbin.setLocator(vn, gl.ELoc.Z)

    # Create neighborhood
    neigh = _create_neigh(neighborhood)

    # Handle drifts
    if ext_drift is not None or pol_drift is not None:
        # Universal kriging
        model_copy = model.clone()

        if pol_drift is not None:
            model_copy.setDriftIRF(pol_drift)

        if ext_drift is not None:
            if isinstance(ext_drift, str):
                ext_drift = [ext_drift]

            for drift in ext_drift:
                dbin.setLocator(drift, gl.ELoc.F)

            for _ in ext_drift:
                model_copy.addDrift(gl.DriftM())

        # Perform cross-validation
        err = gl.xvalid(
            dbin,
            model_copy,
            neigh,
            flag_xvalid_est=True,
            flag_xvalid_std=std,
            namconv=gl.NamingConvention(prefix),
        )

    elif type == "simple":
        # Simple kriging
        if mean is None:
            raise ValueError("Mean must be specified for simple kriging.")

        model_copy = model.clone()
        model_copy.setMean(mean)

        err = gl.xvalid(
            dbin,
            model_copy,
            neigh,
            flag_xvalid_est=True,
            flag_xvalid_std=std,
            namconv=gl.NamingConvention(prefix),
        )

    else:
        # Ordinary kriging
        model_copy = model.clone()
        model_copy.addDrift(gl.DriftM())

        err = gl.xvalid(
            dbin,
            model_copy,
            neigh,
            flag_xvalid_est=True,
            flag_xvalid_std=std,
            namconv=gl.NamingConvention(prefix),
        )


def kriging_mean(db, vname, model):
    """
    Calculate the kriging mean of a variable.

    Args:
        db: gstlearn Db object
        vname: Variable name
        model: gstlearn Model object

    Returns:
        Float value representing the kriging mean

    Examples:
        >>> import minigst as mg
        >>> km = mg.kriging_mean(db, vname='elevation', model=model)
    """
    # Set variable
    db.setLocator(vname, gl.ELoc.Z)

    # Create selection for non-NA values
    values = db[vname]
    sel = ~np.isnan(values)
    db["_sel_temp"] = sel.astype(float)
    db.setLocator("_sel_temp", gl.ELoc.SEL)

    # # Create target at data points
    # coord_names = db.getNamesByLocator(gl.ELoc.X)
    # coords = []
    # for i, name in enumerate(coord_names):
    #     coords.append(db.getCenters(useSel=True)[i])

    # # Create a simple Db for targets
    # tgt = create_dbgrid(coords = coords, coord_names=coord_names)
    # tgt.setLocators(coord_names, gl.ELoc.X)

    df = pd.DataFrame(
        {"x": [db.getCenters(useSel=True)[0]], "y": [db.getCenters(useSel=True)[1]]}
    )
    df = pd.concat([df, df])
    tgt = df_to_db(df, coord_names=["x", "y"])

    tgt.display()
    # Unique neighborhood
    neighU = gl.NeighUnique.create()

    # Simple Kriging with mean 0
    mod0 = model.clone()
    mod0.setMean(0.0)
    err = gl.kriging(
        db,
        tgt,
        mod0,
        neighU,
        flag_est=True,
        flag_std=False,
        namconv=gl.NamingConvention("SK0"),
    )

    # Simple Kriging with mean 1
    mod1 = model.clone()
    mod1.setMean(1.0)
    err = gl.kriging(
        db,
        tgt,
        mod1,
        neighU,
        flag_est=True,
        flag_std=False,
        namconv=gl.NamingConvention("SK1"),
    )

    # Ordinary Kriging
    modOK = model.clone()
    modOK.addDrift(gl.DriftM())
    err = gl.kriging(
        db,
        tgt,
        modOK,
        neighU,
        flag_est=True,
        flag_std=False,
        namconv=gl.NamingConvention("OK"),
    )

    # Calculate kriging mean
    ok_vals = tgt[tgt.getNamesByLocator(gl.ELoc.Z)[0]]
    sk0_vals = tgt["SK0." + vname + ".estim"]
    sk1_vals = tgt["SK1." + vname + ".estim"]

    km = np.mean((ok_vals - sk0_vals) / (sk1_vals - sk0_vals))

    # Cleanup
    db.deleteColumn("_sel_temp")

    return km


def set_mean(model, mu):
    """
    Set the mean value in a Model object.

    Args:
        model: gstlearn Model object
        mu: Mean value to set

    Examples:
        >>> import minigst as mg
        >>> mg.set_mean(model, 10.5)
    """
    model.setMean(mu)


# def kriging_mean(db, vname, model):
#     """
#     Calculate the kriging mean of a variable.

#     Args:
#         db: gstlearn Db object
#         vname: Variable name
#         model: gstlearn Model object

#     Returns:
#         Float value representing the kriging mean

#     Examples:
#         >>> import minigst as mg
#         >>> km = mg.kriging_mean(db, vname='elevation', model=model)
#     """
#     # Set variable
#     db.setLocator(vname, gl.ELoc.Z)

#     # Create selection for non-NA values
#     values = db[vname]
#     sel = ~np.isnan(values)
#     db["_sel_temp"] = sel.astype(float)
#     db.setLocator("_sel_temp", gl.ELoc.SEL)

#     # Create target at data points
#     coords = []
#     coord_names = db.getNamesByLocator(gl.ELoc.X)
#     for name in coord_names:
#         coords.append(db[name][sel])

#     # Create a simple Db for targets
#     tgt = gl.Db()
#     for i, name in enumerate(coord_names):
#         tgt[name] = coords[i]
#     tgt.setLocators(coord_names, gl.ELoc.X)

#     # Unique neighborhood
#     neighU = gl.NeighUnique.create()

#     # Simple Kriging with mean 0
#     mod0 = model.clone()
#     mod0.setMean(0.0)
#     err = gl.kriging(db, tgt, mod0, neighU, flag_est=True, flag_std=False,
#                     namconv=gl.NamingConvention("SK0"))

#     # Simple Kriging with mean 1
#     mod1 = model.clone()
#     mod1.setMean(1.0)
#     err = gl.kriging(db, tgt, mod1, neighU, flag_est=True, flag_std=False,
#                     namconv=gl.NamingConvention("SK1"))

#     # Ordinary Kriging
#     modOK = model.clone()
#     modOK.addDrift(gl.DriftM())
#     err = gl.kriging(db, tgt, modOK, neighU, flag_est=True, flag_std=False,
#                     namconv=gl.NamingConvention("OK"))

#     # Calculate kriging mean
#     ok_vals = tgt[tgt.getNamesByLocator(gl.ELoc.Z)[0]]
#     sk0_vals = tgt["SK0." + vname + ".estim"]
#     sk1_vals = tgt["SK1." + vname + ".estim"]

#     km = np.mean((ok_vals - sk0_vals) / (sk1_vals - sk0_vals))

#     # Cleanup
#     db.deleteColumn("_sel_temp")

#     return km


# def set_mean(model, mu):
#     """
#     Set the mean value in a Model object.

#     Args:
#         model: gstlearn Model object
#         mu: Mean value to set

#     Examples:
#         >>> import minigst as mg
#         >>> mg.set_mean(model, 10.5)
#     """
#     model.setMean(mu)
